crates/circuit/src/parameter_table.rs

```diff
@@ -90,55 +72,24 @@ impl ParameterUuid {
             symbol.uuid.as_u128()
         } else {
             return Err(PyTypeError::new_err(
-                "Could not downcast to Parameter or Expression",
+                "Could not downcast to Parameter or Expression (that equals a symbol)",
             ));
         };
 
         Ok(Self(uuid))
     }
-}
-
-/// This implementation of `FromPyObject` is for the UUID itself, which is what the `ParameterUuid`
-/// struct actually encapsulates.
-impl<'py> FromPyObject<'py> for ParameterUuid {
-    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
-        if ob.is_exact_instance(UUID.get_bound(ob.py())) {
-            ob.getattr(intern!(ob.py(), "int"))?.extract().map(Self)
-        } else {
-            Err(PyTypeError::new_err("not a UUID"))
-        }
-    }
-}
 
-/// Type-safe UUID for a parameter vector.  This is just used internally for tracking.
-#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
-struct VectorUuid(u128);
-impl VectorUuid {
-    /// Extract a UUID from a Python-space `ParameterVector` object. This assumes that the object is
-    /// the correct type.
-    fn from_vector(ob: &Bound<PyAny>) -> PyResult<Self> {
-        ob.getattr(intern!(ob.py(), "_root_uuid"))?.extract()
-    }
-}
-impl<'py> FromPyObject<'py> for VectorUuid {
-    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
-        if ob.is_exact_instance(UUID.get_bound(ob.py())) {
-            ob.getattr(intern!(ob.py(), "int"))?.extract().map(Self)
-        } else {
-            Err(PyTypeError::new_err("not a UUID"))
-        }
+    pub fn from_symbol(symbol: &Symbol) -> Self {
+        Self(symbol.uuid.as_u128())
     }
 }
 
 #[derive(Clone, Default, Debug)]
 pub struct ParameterTable {
```

Comment:
This is a future thing since I think it's probably too much for this PR, but I think we can make the parameter table store `&Symbol` directly with a lifetime that gets associated to the owning circuit. I don't think the table shouldn't need to own any data about the parameters here because they already exist in the circuit and the parameter necessarily has to live as long here as it exists in the circuit. But for this first round just getting use to using an expression directly in `Param` is the main intent of this PR we can optimize the parameter table further after this migration is made.